pkg/manifestclient/write_roundtripper.go

```diff
@@ -0,0 +1,133 @@
+		requestInfoResolver: server.NewRequestInfoResolver(&server.Config{
+			LegacyAPIGroupPrefixes: sets.NewString(server.DefaultLegacyAPIPrefix),
+		}),
+	}
+}
+
+func (mrt *writeTrackingRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	resp := &http.Response{}
+
+	retJSONBytes, err := mrt.roundTrip(req)
+	if err != nil {
+		resp.StatusCode = http.StatusInternalServerError
+		resp.Status = http.StatusText(resp.StatusCode)
+		resp.Body = io.NopCloser(bytes.NewBufferString(err.Error()))
+		return resp, nil
+	}
+
+	resp.StatusCode = http.StatusOK
+	resp.Status = http.StatusText(resp.StatusCode)
+	resp.Body = io.NopCloser(bytes.NewReader(retJSONBytes))
+	// We always return application/json. Avoid clients expecting proto for built-ins.
+	// this may or may not work for apply.  Guess we'll find out.
+	resp.Header = make(http.Header)
+	resp.Header.Set("Content-Type", "application/json")
+
+	return resp, nil
+}
+
+func (mrt *writeTrackingRoundTripper) roundTrip(req *http.Request) ([]byte, error) {
+	requestInfo, err := mrt.requestInfoResolver.NewRequestInfo(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed reading requestInfo: %w", err)
+	}
+
+	if !requestInfo.IsResourceRequest {
+		return nil, fmt.Errorf("non-resource requests are not supported by this implementation")
+	}
+	if len(requestInfo.Subresource) != 0 && requestInfo.Subresource != "status" {
+		return nil, fmt.Errorf("subresource %v is not supported by this implementation", requestInfo.Subresource)
+	}
+
+	var action Action
+	switch {
+	case requestInfo.Verb == "create" && len(requestInfo.Subresource) == 0:
+		action = ActionCreate
+	case requestInfo.Verb == "update" && len(requestInfo.Subresource) == 0:
+		action = ActionUpdate
+	case requestInfo.Verb == "update" && requestInfo.Subresource == "status":
+		action = ActionUpdateStatus
+	case requestInfo.Verb == "patch" && req.Header.Get("Content-Type") == string(types.ApplyPatchType) && len(requestInfo.Subresource) == 0:
+		action = ActionApply
+	case requestInfo.Verb == "patch" && req.Header.Get("Content-Type") == string(types.ApplyPatchType) && requestInfo.Subresource == "status":
+		action = ActionApplyStatus
+	case requestInfo.Verb == "delete" && len(requestInfo.Subresource) == 0:
+		action = ActionDelete
+	default:
+		return nil, fmt.Errorf("verb %v is not supported by this implementation", requestInfo.Verb)
+	}
+
+	bodyContent := []byte{}
+	if req.Body != nil {
+		bodyContent, err = io.ReadAll(req.Body)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read body: %w", err)
+		}
+	}
+	bodyObj, err := runtime.Decode(unstructured.UnstructuredJSONScheme, bodyContent)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode body: %w", err)
+	}
+	bodyYAMLBytes, err := yaml.Marshal(bodyObj.(*unstructured.Unstructured).Object)
+	if err != nil {
+		return nil, fmt.Errorf("unable to encode body: %w", err)
+	}
+	bodyJSONBytes, err := json.Marshal(bodyObj.(*unstructured.Unstructured).Object)
```

Comment:
why ? cannot you just return `bodyObj` ?

Reply:
> why ? cannot you just return `bodyObj` ?

roundtrippers return bytes.  if we desire a refactor at some point, maybe.  But I think this gives us flexibility where we want it.